SPROG Simulator mode option

silverailscolo opened the pull request:
Lets user choose the operating mode for SprogSim

bobjacobsen commented:
Travis graphical restarted after (multiple items, note the 2nd mentions SPROG):
```
[INFO] Running jmri.jmrit.display.layoutEditor.LayoutEditorWindowTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.504 s - in jmri.jmrit.display.layoutEditor.LayoutEditorWindowTest
[INFO] Running jmri.jmrit.display.layoutEditor.LayoutEditorTest
WARN  - Cleaning up frame "Change Sensor Icons" (a class javax.swing.JFrame) in jmri.jmrit.display.layoutEditor.LayoutEditorTest. [main] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "Change Signal Head Icons" (a class javax.swing.JFrame) in jmri.jmrit.display.layoutEditor.LayoutEditorTest. [main] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "Change Icon" (a class javax.swing.JFrame) in jmri.jmrit.display.layoutEditor.LayoutEditorTest. [main] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "Change Sensor Icons" (a class javax.swing.JFrame) in jmri.jmrit.display.layoutEditor.LayoutEditorTest. [main] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "Change Signal Head Icons" (a class javax.swing.JFrame) in jmri.jmrit.display.layoutEditor.LayoutEditorTest. [main] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "Change Icon" (a class javax.swing.JFrame) in jmri.jmrit.display.layoutEditor.LayoutEditorTest. [main] jmri.util.JUnitUtil.resetWindows()
[ERROR] Tests run: 96, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 203.818 s <<< FAILURE! - in jmri.jmrit.display.layoutEditor.LayoutEditorTest
[ERROR] testSetHighlightSelectedBlockFalse(jmri.jmrit.display.layoutEditor.LayoutEditorTest)  Time elapsed: 180.488 s  <<< ERROR!
org.netbeans.jemmy.TimeoutExpiredException: Wait event queue staying empty for 100 (QueueTool.WaitQueueEmptyTimeout)
	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:213)
	at org.netbeans.jemmy.QueueTool.waitEmpty(QueueTool.java:294)
	at jmri.jmrit.display.layoutEditor.LayoutEditorTest.testSetHighlightSelectedBlockFalse(LayoutEditorTest.java:720)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)

[INFO] Running jmri.jmrit.display.layoutEditor.TrackSegmentTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.384 s - in jmri.jmrit.display.layoutEditor.TrackSegmentTest

```

```
[INFO] Running apps.SoundPro.BundleTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in apps.SoundPro.BundleTest
[INFO] Running apps.SoundPro.SoundProTest
INFO  - Starting with profile LocoNet_Simulator_Empty.3f01644a [Time-limited test] apps.Apps.<init>()
WARN  - Cleaning up frame "LocoNet Simulator" (a class jmri.jmrix.loconet.hexfile.HexFileFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "JMRI System console" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "rotate text test" (a class jmri.jmrit.display.layoutEditor.LayoutEditor) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "A" (a class jmri.jmrit.display.layoutEditor.LayoutEditor) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "B" (a class jmri.jmrit.display.layoutEditor.LayoutEditor) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "SoundPro" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
INFO  - Starting with profile TMCC Simulator.3f033c2b [Time-limited test] apps.Apps.<init>()
WARN  - Cleaning up frame "Preferences" (a class apps.gui3.TabbedPreferencesFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "JMRI" (a class apps.SplashWindow) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "SoundPro" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
WARN  - Cleaning up frame "SoundPro" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
INFO  - Starting with profile LocoNet_Simulator.3eac0cdc [Time-limited test] apps.Apps.<init>()
WARN  - Cleaning up frame "SoundPro" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
INFO  - Starting with profile EasyDcc_Simulator.3f033c2b [Time-limited test] apps.Apps.<init>()
WARN  - Cleaning up frame "SoundPro" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.resetWindows()
[ERROR] Tests run: 5, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 120.685 s <<< FAILURE! - in apps.SoundPro.SoundProTest
[ERROR] testLaunchSprog(apps.SoundPro.SoundProTest)  Time elapsed: 90.011 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 90 seconds
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:257)
	at java.awt.Dialog.show(Dialog.java:1084)
	at javax.swing.JOptionPane.showOptionDialog(JOptionPane.java:869)
	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:666)
	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:637)
	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:204)
	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:170)
	at apps.Apps.<init>(Apps.java:252)
	at apps.SoundPro.SoundPro.<init>(SoundPro.java:41)
	at apps.SoundPro.SoundPro.main(SoundPro.java:109)
	at apps.SoundPro.SoundProTest.testLaunchSprog(SoundProTest.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)

[INFO] Running apps.SplashWindowTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 s - in apps.SplashWindowTest
```

bobjacobsen commented:
AppVeyor restarted after 
```
junit] ------------- ---------------- ---------------
    [junit] Testcase: testLaunchSprog(apps.SoundPro.SoundProTest):	Caused an ERROR
    [junit] test timed out after 90 seconds
    [junit] org.junit.runners.model.TestTimedOutException: test timed out after 90 seconds
    [junit] 	at java.lang.Object.wait(Native Method)
    [junit] 	at java.lang.Object.wait(Object.java:502)
    [junit] 	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:264)
    [junit] 	at java.awt.Dialog.show(Dialog.java:1084)
    [junit] 	at javax.swing.JOptionPane.showOptionDialog(JOptionPane.java:869)
    [junit] 	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:666)
    [junit] 	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:637)
    [junit] 	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:204)
    [junit] 	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:170)
    [junit] 	at apps.Apps.<init>(Apps.java:252)
    [junit] 	at apps.SoundPro.SoundPro.<init>(SoundPro.java:41)
    [junit] 	at apps.SoundPro.SoundPro.main(SoundPro.java:109)
    [junit] 	at apps.SoundPro.SoundProTest.testLaunchSprog(SoundProTest.java:125)
    [junit] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    [junit] 	at java.lang.Thread.run(Thread.java:748)
    [junit] 
    [junit] 
    [junit] ERROR - reply complete in unexpected state: 10 was O
mvn :     [junit] Test apps.tests.AllTest FAILED
```

Note there where were a _lot_ of null pointer exceptions before this.

bobjacobsen commented:
AppVeyor failed on what seems to be SPROG-related fault, which might be related to this change.  Not restarting so the log stays available.

```
   [junit] ------------- ---------------- ---------------
    [junit] Testcase: testLaunchSprog(apps.PanelPro.PanelProTest):	Caused an ERROR
    [junit] Required nonnull default for jmri.jmrix.ConnectionConfigManager does not exist.
    [junit] java.lang.NullPointerException: Required nonnull default for jmri.jmrix.ConnectionConfigManager does not exist.
    [junit] 	at jmri.InstanceManager.getDefault(InstanceManager.java:212)
    [junit] 	at apps.Apps.statusPanel(Apps.java:826)
    [junit] 	at apps.PanelPro.PanelPro.statusPanel(PanelPro.java:68)
    [junit] 	at apps.Apps.<init>(Apps.java:396)
    [junit] 	at apps.PanelPro.PanelPro.<init>(PanelPro.java:40)
    [junit] 	at apps.PanelPro.PanelPro.main(PanelPro.java:102)
    [junit] 	at apps.PanelPro.PanelProTest.testLaunchSprog(PanelProTest.java:136)
    [junit] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    [junit] 	at java.lang.Thread.run(Thread.java:748)
    [junit] 
    [junit] 
    [junit] Testcase: testLaunchSprog(apps.SoundPro.SoundProTest):	Caused an ERROR
    [junit] test timed out after 90 seconds
    [junit] org.junit.runners.model.TestTimedOutException: test timed out after 90 seconds
    [junit] 	at java.lang.Object.wait(Native Method)
    [junit] 	at java.lang.Object.wait(Object.java:502)
    [junit] 	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:264)
    [junit] 	at java.awt.Dialog.show(Dialog.java:1084)
    [junit] 	at javax.swing.JOptionPane.showOptionDialog(JOptionPane.java:869)
    [junit] 	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:666)
    [junit] 	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:637)
    [junit] 	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:204)
    [junit] 	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:170)
    [junit] 	at apps.Apps.<init>(Apps.java:252)
    [junit] 	at apps.SoundPro.SoundPro.<init>(SoundPro.java:41)
    [junit] 	at apps.SoundPro.SoundPro.main(SoundPro.java:109)
    [junit] 	at apps.SoundPro.SoundProTest.testLaunchSprog(SoundProTest.java:125)
    [junit] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    [junit] 	at java.lang.Thread.run(Thread.java:748)
    [junit] 
    [junit] 
    [junit] ERROR - reply complete in unexpected state: 10 was O
```

silverailscolo commented:
Thanks @bobjacobsen. Not essential for upcoming release.
apps.SoundPro.SoundProTest again crashes the JVM when run locally so I'm unshure where to look:
```
libc++abi.dylib: terminating with uncaught exception of type NSException
Testsuite: apps.SoundPro.SoundProTest
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0 sec

Testcase: apps.SoundPro.SoundProTest:testLaunchSprog:	Caused an ERROR
Forked Java VM exited abnormally. Please note the time in the report does not reflect the time until the VM exit.
junit.framework.AssertionFailedError: Forked Java VM exited abnormally. Please note the time in the report does not reflect the time until the VM exit.
	at com.intellij.rt.ant.execution.AntMain2.main(AntMain2.java:30)


Running apps.SoundPro.SoundProTest
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0 sec
Test apps.SoundPro.SoundProTest FAILED (crashed)
/Users/egbertbroerse/Documents/Egbert/Computers/IntelliJ local/JMRI/build.xml:685: condition satisfied
	at org.apache.tools.ant.taskdefs.Exit.execute(Exit.java:164)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
	at org.apache.tools.ant.Main.runBuild(Main.java:851)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.Main.start(Main.java:198)
	at org.apache.tools.ant.Main.main(Main.java:286)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.ant.execution.AntMain2.main(AntMain2.java:30)
/Users/egbertbroerse/Documents/Egbert/Computers/IntelliJ local/JMRI/build.xml:685: condition satisfied
	at org.apache.tools.ant.taskdefs.Exit.execute(Exit.java:164)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
	at org.apache.tools.ant.Main.runBuild(Main.java:851)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.Main.start(Main.java:198)
	at org.apache.tools.ant.Main.main(Main.java:286)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.ant.execution.AntMain2.main(AntMain2.java:30)
/Users/egbertbroerse/Documents/Egbert/Computers/IntelliJ local/JMRI/build.xml:685: condition satisfied
	at org.apache.tools.ant.taskdefs.Exit.execute(Exit.java:164)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
	at org.apache.tools.ant.Main.runBuild(Main.java:851)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.Main.start(Main.java:198)
	at org.apache.tools.ant.Main.main(Main.java:286)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.ant.execution.AntMain2.main(AntMain2.java:30)
```

bobjacobsen commented:
I'm also seeing a problem running locally:
```
JUnit version 4.12
.Using org.netbeans.jemmy.drivers.APIDriverInstaller driver installer
INFO  - Starting with profile My_JMRI_Railroad.3f323e91 [Time-limited test] apps.Apps.?()
INFO  - No saved preferences, will open preferences window.  Searched for /Users/jake/Documents/Trains/JMRI/projects/JMRI/temp/My_JMRI_Railroad/ProfileConfig.xml [Time-limited test] apps.Apps.?()
WARN  - Cleaning up frame "Preferences" (a class apps.gui3.TabbedPreferencesFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.?()
WARN  - Cleaning up frame "SoundPro" (a class jmri.util.JmriJFrame) in apps.SoundPro.SoundProTest. [Time-limited test] jmri.util.JUnitUtil.?()
.#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff30a891c4, pid=78664, tid=0x0000000000017003
#
# JRE version: Java(TM) SE Runtime Environment (8.0_151-b12) (build 1.8.0_151-b12)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.151-b12 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [CoreFoundation+0x531c4]  CFSetGetValue+0x24
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/jake/Documents/Trains/JMRI/projects/JMRI/hs_err_pid78664.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
./.run.sh: line 550: 78664 Abort trap: 6           "${JAVACMD}" "${DOCK_OPTIONS[@]}" ${OPTIONS} ${ALTPORTS} ${CONFIGFILE} -cp "${CP}" "${CLASSNAME}" ${ARGS}
```

The dump shows the problem is on:
```
Current thread (0x00007ff8f59a2800):  JavaThread "hid4java Device Scanner" daemon [_thread_in_native, id=94211, stack(0x000070000f18e000,0x000070000f28e000)]
```

The Java traceback for that is:
```
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.sun.jna.Native.invokePointer(Lcom/sun/jna/Function;JI[Ljava/lang/Object;)J+0
j  com.sun.jna.Function.invokePointer(I[Ljava/lang/Object;)Lcom/sun/jna/Pointer;+7
j  com.sun.jna.Function.invoke([Ljava/lang/Object;Ljava/lang/Class;ZI)Ljava/lang/Object;+440
j  com.sun.jna.Function.invoke(Ljava/lang/reflect/Method;[Ljava/lang/Class;Ljava/lang/Class;[Ljava/lang/Object;Ljava/util/Map;)Ljava/lang/Object;+271
j  com.sun.jna.Library$Handler.invoke(Ljava/lang/Object;Ljava/lang/reflect/Method;[Ljava/lang/Object;)Ljava/lang/Object;+344
j  com.sun.proxy.$Proxy7.hid_enumerate(SS)Lorg/hid4java/jna/HidDeviceInfoStructure;+26
j  org.hid4java.jna.HidApi.enumerateDevices(II)Lorg/hid4java/jna/HidDeviceInfoStructure;+7
j  org.hid4java.HidDeviceManager.getAttachedHidDevices()Ljava/util/List;+10
j  org.hid4java.HidDeviceManager.scan()V+9
j  org.hid4java.HidDeviceManager$3.run()V+43
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
```

What may be happening:
- There was a recent addition of support for HID (USB) devices that starts some HID library processing at startup. (That's almost certainly a mistake, as the vast majority of JMRI instances don't do anything with HID devices per se, but even poor code wins sometimes)
- Since we're starting up multiple app instances in test, perhaps we're either not stopping the HID library support properly when ending the app (even poor code wins sometimes), or we're tripping a bug in the library by stopping and restarting it.

More to follow.

bobjacobsen commented:
I'm working on the crash in the SoundPro test and will have a PR shortly.

In the meantime, I've got it mostly running, which lets me see why the SPROG test is failing: It's popping this window and then stalling:
![error_initializing_soundpro_and_jmri](https://user-images.githubusercontent.com/2774117/35187363-a3bce94c-fdd7-11e7-82bf-7a285ea94048.png)

Can you try to fix that in the profile the test is running?  Thanks.

silverailscolo commented:
I had changed Simulator mode in Ctor to start off as Command Station (from Programmer). Not exactly sure why, but that seems to cause the null ThrottleManager. Reverted, works locally. Let's see if CI agrees.

silverailscolo commented:
Travis Graphic error:
```
[INFO] Running jmri.jmrit.logix.MergePromptTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 10.014 s <<< FAILURE! - in jmri.jmrit.logix.MergePromptTest
[ERROR] testCTor(jmri.jmrit.logix.MergePromptTest)  Time elapsed: 10.014 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:257)
	at java.awt.Dialog.show(Dialog.java:1084)
	at java.awt.Component.show(Component.java:1671)
	at java.awt.Component.setVisible(Component.java:1623)
	at java.awt.Window.setVisible(Window.java:1014)
	at java.awt.Dialog.setVisible(Dialog.java:1005)
	at jmri.jmrit.logix.MergePrompt.<init>(MergePrompt.java:148)
	at jmri.jmrit.logix.MergePromptTest.testCTor(MergePromptTest.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

silverailscolo commented:
Here's the problem: As this new options seeks to change the capabilities of the connection in the SystemConnectionMemo after initialization (using the existing method ``memo.setSprogMode(OPS)``), its Default duties have already been used to take over from Internal. So it is assigned a task like Throttle it cannot fulfill. With the Throttle reset to Default, the Command Station mode gives no more errors (but you can't see any traffic in the SROG Console).
I's think it is best to take some time to get advice from @rhwood and not put this in yet.
![sprogsimdefaultthrottle](https://user-images.githubusercontent.com/14811327/35188524-b9edcd80-fe36-11e7-819d-6f0d514ac0b1.png)

rhwood commented:
@silverailscolo How should I be recreating this failure state?

silverailscolo commented:
Add a SPROG Simulator connection, set the Additional Connection Settings > Operate as: to "Command Station". If you set it to operate as a "SPROG Programmer", no problem (BTW it's the other way around in current master/4.11.2 test).
![sprogsimmode](https://user-images.githubusercontent.com/14811327/35192285-0b14da70-fe8f-11e7-88ba-1f1e14e3edac.png)
Look at the Prefs Defaults tab (shown above): if the SPROG Sim is selected as default for Throttle, after restart it will show the initialization error message (shown above).

silverailscolo commented:
For 4.11.2-extra. Commit https://github.com/JMRI/JMRI/pull/4795/commits/2b5fa454a722c9b9ea754239f9205abea799b08d fixes SPROG Sim config option - null Default Throttle Manager error (option was accidentally merged into master)

rhwood commented:
@silverailscolo I can't reproduce the problem you saw after pulling 0df38ab, so I am assuming you managed to fix it.

Some comments though:
- In CS mode, the log is spammed with messages like `WARN  - Packet delay was 54 ms time now 1516540043851 [SPROG slot thread]`; you probably want them to be debug or trace level messages.
- Is someone likely to want to switch modes? The defaults do not follow the change in SPROG simulator mode, is this expected to be a problem? You might want to provide two simulators (one for each mode) instead of one simulator with a settable mode, since having different simulation profiles might help users with the management of defaults.
- You probably want to remove the default (no parameters) constructor for SprogSystemConnectionMemo, since it can only create an invalid connection (SprogMode does not get set); alternatively, you might want it to call `this(SprogMode.OPS);` so it does create a valid connection (having multiple constructors use `this(...)` with different parameters reduces code duplication).

silverailscolo commented:
Thanks for reviewing @rhwood 
Providing the Simulator with the option I preferred over the pop up getting even longer. Also from a code duplication view I think it keeps the number of different Adapters down. This option being a bit hidden seems not bad; better anyway that me fixing the Sim to either mode.
The debug traffic is being worked on by @crosland but at the moment it indeed is flooding the console so I reduced its level to "trace". It wasn't complaining 2 weeks ago while I was creating this element. By the time the Simulator has been improved to come up with a proper reply in time, the level can be stepped up again.

rhwood commented:
@silverailscolo I think the tests are failing on some assumptions made when the SPROG mode is unset in the SprogSystemConnectionMemo. You might want to adjust the test assumptions with a default constructor for SprogSystemConnectionMemo to match how a SPROG in OPS mode is handled.

silverailscolo commented:
Travis Graphic error seems unrelated:
```
[ERROR] Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 114.404 s <<< FAILURE! - in apps.DispatcherPro.DispatcherProTest
[ERROR] testLaunchTmcc(apps.DispatcherPro.DispatcherProTest)  Time elapsed: 90.037 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 90 seconds
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:257)
	at java.awt.Dialog.show(Dialog.java:1084)
	at javax.swing.JOptionPane.showOptionDialog(JOptionPane.java:869)
	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:666)
	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:637)
	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:204)
	at jmri.implementation.JmriConfigurationManager.load(JmriConfigurationManager.java:170)
	at apps.Apps.<init>(Apps.java:252)
	at apps.DispatcherPro.DispatcherPro.<init>(DispatcherPro.java:40)
	at apps.DispatcherPro.DispatcherPro.main(DispatcherPro.java:102)
	at apps.DispatcherPro.DispatcherProTest.testLaunchTmcc(DispatcherProTest.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

bobjacobsen commented:
Setting aside the test issues (actually running JMRI with layouts is what test releases are about), is this the PR that you want in 4.11.2?

pabender commented:
The last Travis headed run failed on MergePromptTest, which is not related to this PR.  I went ahead and restarted

silverailscolo commented:
@bobjacobsen Yes, this is it.

bobjacobsen commented:
Graphical travis restarted after:
```
[INFO] Running jmri.jmrit.operations.ExceptionContextTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in jmri.jmrit.operations.ExceptionContextTest
[INFO] Running jmri.jmrit.operations.ExceptionDisplayFrameTest
```

bobjacobsen commented:
This is now showing a conflict in SprogCommandStation.  Which one (top or bottom) should I take?
```
                // Get next packet to send
                byte[] p;
                SprogSlot s = sendNow.poll();
                if (s != null) {
                    // New throttle action to be sent immediately
                    p = s.getPayload();
                    log.debug("Packet from immediate send queue");
                } else {
                    // Or take the next one from the stack
                    p = getNextPacket();
                    if (p != null) {
<<<<<<< eb-sprog-simulator
                        // Send the packet
                        sendPacket(p, SprogConstants.S_REPEATS);
                        log.debug("Packet sent");
                    } else {
                        // Send a decoder idle packet to prompt a reply from hardware and keep things running
                        sendPacket(jmri.NmraPacket.idlePacket(), SprogConstants.S_REPEATS);
                    }
                    timeNow = System.currentTimeMillis();
                    packetDelay = timeNow - time;
                    time = timeNow;
                    // Useful for debug if packets are being delayed; Set to trace level to be able to debug other stuff
                    if (packetDelay > MAX_PACKET_DELAY) {
                        log.trace("Packet delay was {} ms time now {}", packetDelay, time);
=======
                        log.debug("Packet from stack");
>>>>>>> master
                    }
                }
                replyAvailable = false;
                if (p != null) {
                    // Send the packet
                    sendPacket(p, SprogConstants.S_REPEATS);
                    log.debug("Packet sent");
                } else {
                    // Send a decoder idle packet to prompt a reply from hardware and keep things running
                    sendPacket(jmri.NmraPacket.idlePacket(), SprogConstants.S_REPEATS);
                }
```

silverailscolo commented:
Used the new CS code, except for the debug level (after a nightly break)

bobjacobsen commented:
Thanks!  Will merge and rerun 4.11.2
